x/mm2/keeper/keeper.go

```diff
@@ -0,0 +1,158 @@
+}
+
+// GetMarket returns a market from the store by its currency pair string ID.
+func (k *Keeper) GetMarket(ctx sdk.Context, tickerStr string) (types.Market, error) {
+	return k.markets.Get(ctx, types.TickerString(tickerStr))
+}
+
+// GetAllMarkets returns the set of Market objects currently stored in state
+// as a map[TickerString] -> Markets.
+func (k *Keeper) GetAllMarkets(ctx sdk.Context) (map[string]types.Market, error) {
+	iter, err := k.markets.Iterate(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	keyValues, err := iter.KeyValues()
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string]types.Market, len(keyValues))
+	for _, keyValue := range keyValues {
+		m[string(keyValue.Key)] = keyValue.Value
+	}
+
+	return m, nil
+}
+
+// CreateMarket initializes a new Market.
+// The Ticker.String corresponds to a market, and must be unique.
+func (k *Keeper) CreateMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if alreadyExists {
+		return types.NewMarketAlreadyExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
+}
+
+// UpdateMarket updates a Market.
+// The Ticker.String corresponds to a market, and exist unique.
+func (k *Keeper) UpdateMarket(ctx sdk.Context, market types.Market) error {
+	// Check if Ticker already exists for the provider
+	alreadyExists, err := k.markets.Has(ctx, types.TickerString(market.Ticker.String()))
+	if err != nil {
+		return err
+	}
+	if !alreadyExists {
+		return types.NewMarketDoesNotExistsError(types.TickerString(market.Ticker.String()))
+	}
+	// Create the config
+	return k.markets.Set(ctx, types.TickerString(market.Ticker.String()), market)
+}
+
+// SetParams sets the x/marketmap module's parameters.
+func (k *Keeper) SetParams(ctx sdk.Context, params types.Params) error {
+	return k.params.Set(ctx, params)
+}
+
+// GetParams returns the x/marketmap module's parameters.
+func (k *Keeper) GetParams(ctx sdk.Context) (types.Params, error) {
+	return k.params.Get(ctx)
+}
+
+// ValidateState is called after keeper modifications have been made to the market map to verify that
+// the aggregate of all updates has led to a valid state.
+func (k *Keeper) ValidateState(ctx sdk.Context, updates []types.Market) error {
+	for _, market := range updates {
+		if err := k.IsMarketValid(ctx, market); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// IsMarketValid checks if a market is valid by statefully checking if each of the currency pairs
+// specified by its provider configs are valid and in state.
+func (k *Keeper) IsMarketValid(ctx sdk.Context, market types.Market) error {
+	// check that all markets already exist in the keeper store:
+	for _, providerConfig := range market.ProviderConfigs {
```

Comment:
can we reuse some of the `ValidateBasic` logic here?

Reply:
Not here.  This is some additional stateful validation that requires us to use the keeper.  We need to verify that for each new market added, if they are Normalizing by another market (normalizeBy), does that normalizeBy exist in state.  Any other stateless validation will already have been done by now by `ValidateBasic()`.

Reply:
This protects us from the case where in state we have:
- BTC/USD
- MOG/USD

and we want to add a new market which is:
- ETH/USD (fed by an ETH/USDT feed, normalized by USDT/USD)

Only at a stateful verification stage, can we verify that the required USDT/USD exists.